### PR TITLE
Hardcode offsets in semi-structured wrappers

### PR DESCRIPTION
**Pull request by miklos1**
This change hardcodes the offset values of the structured dimension in the wrapper.

@wence-: Please have a look at f26939b. I use the whole map in the key instead of just replacing arity with the offset array. This might be too discriminating? However, if it is equivalent to adding the individual components by hand, than I prefer this simpler version.

### CODE REVIEW

**Comment by wence-** [comment]
Please don't use the whole map.  This will completely defeat the garbage collector.

**Comment by wence-** [comment]
I think we're careful to not use anything in the cache key that results in holding on to things they reference (but maybe the implicit_bcs and or vector_index already do so)

File: pyop2/utils.py [hunk 1]
```diff
@@ -68,7 +68,7 @@ def __get__(self, obj, cls):
 def as_tuple(item, type=None, length=None):
     # Empty list if we get passed None
     if item is None:
-        t = []
+        t = ()
```
Comment (kynan) [inline pyop2/utils.py]: oops...

**Comment by miklos1** [comment]
@wence-: Fixed caching.

Still waiting for Travis, but otherwise I think this is ready to go.

File: pyop2/base.py [hunk 2]
```diff
@@ -3856,24 +3856,25 @@ def _cache_key(cls, kernel, itspace, *args, **kwargs):
...
                 else:
                     idx = arg.idx
-                map_arity = arg.map.arity if arg.map else None
+                map_offset = tuplify(arg.map.offset) if arg.map else None
                 if arg._is_dat_view:
                     view_idx = arg.data.index
                 else:
                     view_idx = None
-                key += (arg.data.dim, arg.data.dtype, map_arity,
+                key += (arg.data.dim, arg.data.dtype, map_offset,
```
Comment (wence-) [inline pyop2/base.py]: If the map is not extruded, and therefore `arg.map.offset` is `None` this key no longer treats maps with different arities as different.

Comment (miklos1) [inline pyop2/base.py]: Ugh, right. Sadly nothing tests this in the Firedrake test suite. I'll fix.

**Pull request merged**